fix up spec and javadoc requirements with respect to special parameters (#530)
jakartaee/data: spec/src/main/asciidoc/repository.asciidoc

Refer to the Jakarta Data module JavaDoc section on "Return Types for Repository Methods" for a listing of valid return types for methods using Query by Method Name.

=== Special Parameters for Limits, Sorting, and Pagination

An <<Annotated Query methods,annotated>>, <<Parameter-based automatic query methods,parameter-based>>, or <<Query by Method Name,method name>> query method may have _special parameters_ of type `Limit`, `Order`, `Sort`, or `PageRequest`.
A special parameter controls which query results are returned to the caller of a repository method, or in what order the results are returned:

- a `Limit` allows the query results to be limited to a given range defined in terms of an offset and maximum number of results,
- a `Sort` or `Order` allows the query results to be sorted by a given entity field or list of fields, respectively, and
- a `PageRequest` combines a range with optional sorting criteria, and a parameter of this type must be declared when the repository method returns a `Page` of results, as specified below in <<Offset-based Pagination>>, or a `CursoredPage`, as specified in <<Cursor-based Pagination>>.

> A repository method must throw `UnsupportedOperationException` if it has:
> 
> - more than one parameter of type `PageRequest` or `Limit`,
> - a parameter of type `PageRequest` and a parameter of type `Limit`,
> - a parameter of type `PageRequest` or `Limit`, in combination with the keyword `First`, or
> - more than one parameter of type `Order`.
> 
> Alternatively, a Jakarta Data provider is permitted to reject such a repository method declaration at compile time.
> 
> A repository method must throw `IllegalArgumentException` if it is called with an argument of type `PageRequest` with nonempty sort criteria, and a separate argument or arguments of type `Sort` or `Order`.

Reviewer: IllegalArugmentException was stated for this in the JavaDoc updates earlier in this pull.
```suggestion
A repository method must throw `IllegalArugmentException` if it has:

- more than one parameter of type `PageRequest` or `Limit`,
- a parameter of type `PageRequest` and a parameter of type `Limit`,
- a parameter of type `PageRequest` or `Limit`, in combination with the keyword `First`, or
- more than one parameter of type `Order`.
- an argument or arguments of type `Sort` or `Order` and a separate argument of type `PageRequest` with nonempty sort criteria.

Alternatively, a Jakarta Data provider is permitted to reject such a repository method declaration at compile time.
```

Author: I believe the list here is consistent with the list in the Javadoc for `PageRequest`.

Please compare:

```java
 * <p>A repository method may not be declared with:</p>
 * <ul>
 * <li>more than one parameter of type {@code PageRequest} or {@link Limit},
 * <li>a parameter of type {@code PageRequest} and a parameter of type
 *     {@link Limit}, or</li>
 * <li>a parameter of type {@code PageRequest} in combination with the
 *     keyword {@code First}.</li>
 * </ul>
 *
 * <p>A repository method throws {@link IllegalArgumentException} if it
 * is called with an argument of type {@code PageRequest} with nonempty
 * sort criteria, and a separate argument or arguments of type {@link Sort}
 * or {@link Order}.</p>
```

```java
A repository method must throw `UnsupportedOperationException` if it has:

- more than one parameter of type `PageRequest` or `Limit`,
- a parameter of type `PageRequest` and a parameter of type `Limit`,
- a parameter of type `PageRequest` or `Limit`, in combination with the keyword `First`, or
- more than one parameter of type `Order`.

Alternatively, a Jakarta Data provider is permitted to reject such a repository method declaration at compile time.

A repository method must throw `IllegalArgumentException` if it is called with an argument of type `PageRequest` with nonempty sort criteria, and a separate argument or arguments of type `Sort` or `Order`.
```

I believe they are saying the same thing.

> A repository method must throw `UnsupportedOperationException` if the database is incapable of ordering the query results using the given sort criteria.

Reviewer: I missed this one. Refer to discussion in the earlier review comment
```suggestion
A repository method must throw `DataException` if the database is incapable of ordering the query results using the given sort criteria.
```

=== Special Parameters for Limits, Sorting, and Pagination

An <<Annotated Query methods,annotated>>, <<Parameter-based automatic query methods,parameter-based>>, or <<Query by Method Name,method name>> query method may have _special parameters_ of type `Limit`, `Order`, `Sort`, or `PageRequest` if the method return type indicates that the method may return multiple entities, that is, if the return type is:

- an array type,
- `List` or `Stream`, or
- `Page` or `CursoredPage`.

> 

Reviewer: ```suggestion
An <<Annotated Query methods,annotated>>, <<Parameter-based automatic query methods,parameter-based>>, or <<Query by Method Name,method name>> query method that performs a find operation may have _special parameters_ of type `Limit`, `Order`, `Sort`, or `PageRequest`.
```
